synapse/federation/federation_server.py

```diff
@@ -148,6 +148,52 @@ def __init__(self, hs: "HomeServer"):
 
         self._room_prejoin_state_types = hs.config.api.room_prejoin_state
 
+        # Whether we have started handling old events in the staging area.
+        self._started_handling_of_staged_events = False
+
+    def _start_handling_old_staged_events(self) -> None:
+        """Start handling old events in the staging area. This should be called
+        when we receive a new incoming transaction (and are thus an instance
+        that handles incoming federation).
+        """
+        if self._started_handling_of_staged_events:
+            return
+
+        self._started_handling_of_staged_events = True
+
+        self._handle_old_staged_events()
+
+    @wrap_as_background_process("_handle_old_staged_events")
+    async def _handle_old_staged_events(self) -> None:
+        """Handle old staged events by fetching all rooms that have staged
+        events and start the processing of each of those rooms.
+        """
+
+        # Get all the rooms IDs with staged events.
+        room_ids = await self.store.get_all_staged_rooms()
+
+        # We then shuffle them so that if there are multiple instances doing
+        # this work they're less likely to collide.
+        random.shuffle(room_ids)
+
+        for room_id in room_ids:
+            room_version = await self.store.get_room_version(room_id)
+
+            # Try and acquire the processing lock for the room, if we get it start a
+            # background process for handling the events in the room.
+            lock = await self.store.try_acquire_lock(
+                _INBOUND_EVENT_HANDLING_LOCK_NAME, room_id
+            )
+            if lock:
+                self._process_incoming_pdus_in_room_inner(
```

Comment:
probably worth logging something here.

Reply:
I think the common case is that we get the lock, so I wonder if it'd be better to log when we *don't* acquire the lock? (From reducing our log lines perspective?)

Reply:
I've done that for now, but I'm not particularly happy about the wording.

Reply:
(Also note we log `handling received PDU` when we come round to actually handling an event)

Reply:
hrmmm. given this should only happen once (per room with staged events) at startup -  I don't feel like the logging would be too onerous?

Reply:
I completely misread which this was. Sorry.